driver/src/main/com/mongodb/reactivestreams/client/internal/GridFSDownloadPublisherImpl.java

```diff
@@ -0,0 +1,214 @@
+        private boolean isCompleted = false;
+        private boolean isTerminated = false;
+        /* protected by `lock` */
+
+        public GridFSDownloadSubscription(final Subscriber<? super ByteBuffer> outerSubscriber) {
+            this.outerSubscriber = outerSubscriber;
+        }
+
+        private final Subscriber<GridFSFile> gridFSFileSubscriber = new Subscriber<GridFSFile>() {
+            @Override
+            public void onSubscribe(final Subscription s) {
+                s.request(1);
+            }
+
+            @Override
+            public void onNext(final GridFSFile result) {
+                synchronized (lock) {
+                    gridFSFile = result;
+                }
+            }
+
+            @Override
+            public void onError(final Throwable t) {
+                outerSubscriber.onError(t);
+                terminate();
+            }
+
+            @Override
+            public void onComplete() {
+                synchronized (lock) {
+                    requestedData = false;
+                }
+                requestMoreOrComplete();
+            }
+        };
+
+        class GridFSDownloadStreamSubscriber implements Subscriber<Integer> {
+            private final ByteBuffer byteBuffer;
+
+            GridFSDownloadStreamSubscriber(final ByteBuffer byteBuffer) {
+                this.byteBuffer = byteBuffer;
+            }
+
+            @Override
+            public void onSubscribe(final Subscription s) {
+                s.request(1);
+            }
+
+            @Override
+            public void onNext(final Integer integer) {
+                synchronized (lock) {
+                    sizeRead += integer;
+                }
+            }
+
+            @Override
+            public void onError(final Throwable t) {
+                terminate();
+                outerSubscriber.onError(t);
+            }
+
+            @Override
+            public void onComplete() {
+                if (byteBuffer.remaining() > 0) {
+                    gridFSDownloadStream.read(byteBuffer).subscribe(new GridFSDownloadStreamSubscriber(byteBuffer));
+                } else {
+                    synchronized (lock) {
+                        requestedData = false;
+                        if (sizeRead == gridFSFile.getLength()) {
+                            isCompleted = true;
+                        }
+                    }
+                    byteBuffer.flip();
+                    outerSubscriber.onNext(byteBuffer);
+                    requestMoreOrComplete();
+                }
+            }
+        }
+
+        @Override
+        public void request(final long n) {
+            synchronized (lock) {
+                requested += n;
+            }
+            requestMoreOrComplete();
+        }
+
+        @Override
+        public void cancel() {
+            terminate();
+        }
+
+        private void requestMoreOrComplete() {
+            synchronized (lock) {
```

Comment:
Self-recursion caused a `StackOverflowError` in Spring Data especially with small chunk sizes for a large file.

Reply:
I also hit that when running our test suite. I worked around that by, setting the default minimum byte buffer size to be chunk size. Then by handling the batch size in the cursor we can ensure that we don't cause a stackoverflow.